### ISSUE TITLE
Feature request: read_excel to support hyperlinks split into label and hyperlink

### ISSUE THREAD

**Issue by DocViper666** [pandas-dev/pandas#13439]
Situation:
.xlsx file having a column with label and hyperlinks (see screenshot collage)
In Excel, label: Python (that has a hyperlink to i.e.: www.python.org)
#### Code Sample, a copy-pastable example if possible

data = pd.read_excel('Book1.xlsx')
column has just "Python" as content. Hyperlink information is lost.
(see screenshot)

![excel-hyperlink](https://cloud.githubusercontent.com/assets/19929980/16044286/25e27eca-3244-11e6-8ce4-2ae0ec34ba77.png)
#### Expected Output, Proposal:

Add an argument  for example named hyperlinkparser to read_excel()
that offers options: "label", "hyperlink", "both"
label ... just the label (this is how it is handled now)
hyperlink ... just the hyperlink, drop the label
split ... created two separate columns for label and hyperlink each

with example it could look like:
         | MyLinks | MyLinks_hyperlink
0       | Python   | www.python.org

btw. openpyxl 2.4.+ has this feature now included (was broken in previous versions). Maybe this helps for implementation.

Thank you!

**Comment by TomAugspurger** [comment]
How difficult is to preprocess the file with xlrd or openpyxl, and pass the modified version to `read_excel`.? I haven't used either of those libraries extensively. If it's not too much code we could include it as a cookbook example.

**Comment by DocViper666** [comment]
Thanks for the answer. 

What I have found as a workaround 
(and please excuse clumsy coding style - I'm a python beginner):

```
# requires openpyxl version 2.4.0+

from openpyxl import load_workbook
wb = load_workbook(filename='input.xlsx')
ws = wb.get_active_sheet() # ws is now an IterableWorksheet

# iterate thru all cells and if hyperlink found attempt modification of cell
for row in ws.rows:
    for cell in row:
        try:
            if len(cell.hyperlink.target)  > 0:
                cell.value = str(cell.value) + "|"+ str(cell.hyperlink.target) 
                # Join cell.value and hyperlink target into string (optionally just assign the hyperlink.target to the cell.value
        except:
            pass

# save workbook to temp .xlsx (I could not manage to read from buffer...) .
wb.save("temp.xlsx")

# read with pandas and spit+assign to new column
data = pd.read_excel("temp.xlsx")
data = data.assign(Hyperlink = data.MyLinks[0].split("|")[-1])
data = data.assign(LinkLabel = data.MyLinks[0].split("|")[0])

```
Maybe this is helpful for somebody.
An integrated pandas function still would be nice.

Br,
Joe

**Comment by sinhrks** [comment]
One possible enhancement is to allow user funcs (passed via ``converters`` kwds) to accept excel ``Cell`` instances.

**Comment by DocViper666** [comment]
@sinhrks:
Thanks for the feedback. Would this allow to target the cell.value and also the cell.hyperlink.target ? Is this supported in the xlrd module that is used for reading the file?

@all: I have updated my workaround in case someone is looking for ideas.

**Comment by sinhrks** [comment]
@DocViper666 The enhancement should allow something like:

```
pd.read_excel("xxx.xlsx", converters={"target_col_name": lambda x: str(x.value) + "|"+ str(x.hyperlink.target), raw=True)
``` 

As long as I know, xlrd also has a class corresponding to cell.

- http://www.lexicon.net/sjmachin/xlrd.html#xlrd.Cell-class

**Comment by dustinmichels** [comment]
Is anyone working on this? Are pandas maintainers open to pull requests? I would like to work on this, with a little bit of guidance.

**Comment by WillAyd** [comment]
Not aware of anyone working this. If I'm being honest I'm not entirely sure this is within the scope of pandas as it seems like a very niche use, but you are welcome to submit a PR if you find an easy way to implement

**Comment by dustinmichels** [comment]
Okay, thanks for your reply!

I came across a [StackOverflow question](https://stackoverflow.com/questions/35325799/pandas-read-excel-with-hyperlink) about using `pd.read_excel` to read a document with hyperlinks, and the top answer said 'Pandas can't do it, but you can use openpyxl instead.'

It seems in some cases, Pandas already acts as a wrapper around openpyxl. For instance, when I try to use [`pandas.DataFrame.to_excel` ](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.to_excel.html) to save a DataFrame as an Excel file, I get a warning that I should install openpyxl first.

I thought perhaps an option to read an Excel file, maintaining links, could work the same way?

**Comment by WillAyd** [comment]
The openpyxl reader is here:

https://github.com/pandas-dev/pandas/blob/d48306e77c0a708e0ee33a2aa1da7e267df52ef6/pandas/io/excel/_openpyxl.py#L460

`get_sheet_data` would be the method of most interest

**Comment by dustinmichels** [comment]
Thanks!! I'll see what I can do.

**Comment by srihari15** [comment]
Is this being worked on still? Would be very useful if someone could provide a closure. Thanks :)

**Comment by yadalik** [comment]
Quick monkey patching, without converters or anything like this, if you would like to treat ALL cells with hyperlinks as hyperlinks, more sophisticated way, I suppose, at least be able to choose, what columns treat as hyperlinked or gather data, or save somehow both data and hyperlink in same cell at dataframe. And using converters, dunno. (BTW I played also with `data_only`, `keep_links`, did not helped, only changing `read_only` resulted ok, I suppose it can slow down your code speed).

P.S.: Works only with xlsx, i.e.,  engine is openpyxl
P.P.S.: If you reading this comment in the future and issue still Open, don't forget to see changes in `_convert_cell` and `load_workbook` at `pandas.io.excel._openpyxl` and update them accordingly.

```
import pandas
from pandas.io.excel._openpyxl import OpenpyxlReader
import numpy as np
from pandas._typing import FilePathOrBuffer, Scalar


def _convert_cell(self, cell, convert_float: bool) -> Scalar:
    from openpyxl.cell.cell import TYPE_BOOL, TYPE_ERROR, TYPE_NUMERIC
    # here we adding this hyperlink support:
    if cell.hyperlink and cell.hyperlink.target:
        return cell.hyperlink.target
        # just for example, you able to return both value and hyperlink,
        # comment return above and uncomment return below
        # btw this may hurt you on parsing values, if symbols "|||" in value or hyperlink.
        # return f'{cell.value}|||{cell.hyperlink.target}'
    # here starts original code, except for "if" became "elif"
    elif cell.is_date:
        return cell.value
    elif cell.data_type == TYPE_ERROR:
        return np.nan
    elif cell.data_type == TYPE_BOOL:
        return bool(cell.value)
    elif cell.value is None:
        return ""  # compat with xlrd
    elif cell.data_type == TYPE_NUMERIC:
        # GH5394
        if convert_float:
            val = int(cell.value)
            if val == cell.value:
                return val
        else:
            return float(cell.value)

    return cell.value


def load_workbook(self, filepath_or_buffer: FilePathOrBuffer):
    from openpyxl import load_workbook
    # had to change read_only to False:
    return load_workbook(
        filepath_or_buffer, read_only=False, data_only=True, keep_links=False
    )


OpenpyxlReader._convert_cell = _convert_cell
OpenpyxlReader.load_workbook = load_workbook
```

And after adding this above in your python file, you will be able to call `df = pandas.read_excel(input_file)`
After writing all this stuff it came to me, that maybe it would be easier and cleaner just use openpyxl by itself ^_^

**Comment by yadalik** [comment]
Not sure, is it fixed somehow yet, but, this issue still open, so here is updated code for pandas 1.4.0:
```
from __future__ import annotations
import pandas
from pandas.io.excel._openpyxl import OpenpyxlReader
import numpy as np
from pandas._typing import (
    FilePath,
    ReadBuffer,
    Scalar,
)


def _convert_cell(self, cell, convert_float: bool) -> Scalar:
    from openpyxl.cell.cell import (
        TYPE_ERROR,
        TYPE_NUMERIC,
    )
    # here we adding this hyperlink support:
    if cell.hyperlink and cell.hyperlink.target:
        return cell.hyperlink.target
        # just for example, you able to return both value and hyperlink,
        # comment return above and uncomment return below
        # btw this may hurt you on parsing values, if symbols "|||" in value or hyperlink.
        # return f'{cell.value}|||{cell.hyperlink.target}'
    # here starts original code, except for "if" became "elif"
    elif cell.value is None:
        return ""  # compat with xlrd
    elif cell.data_type == TYPE_ERROR:
        return np.nan
    elif not convert_float and cell.data_type == TYPE_NUMERIC:
        return float(cell.value)
    return cell.value


def load_workbook(self, filepath_or_buffer: FilePath | ReadBuffer[bytes]):
    from openpyxl import load_workbook
    # had to change read_only to False (note: may affect on speed of reading file):
    return load_workbook(
        filepath_or_buffer, read_only=False, data_only=True, keep_links=False
    )


OpenpyxlReader._convert_cell = _convert_cell
OpenpyxlReader.load_workbook = load_workbook

# after code above opened Excel file will have hyperlinks:
df = pandas.read_excel(excel_file, sheet_name=None, engine='openpyxl')

```

**Comment by zachschillaci27** [comment]
Going off the other responses, here's a simple solution just using an openpyxl/pandas parser
https://gist.github.com/zachschillaci27/887c272cbdda63316d9a02a770d15040

**Comment by yadalik** [comment]
If google brought you here with error of mine old code, like:
```
    df = pandas.read_excel(input_file, sheet_name=None, engine='openpyxl')
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\python\venv\Lib\site-packages\pandas\io\excel\_base.py", line 495, in read_excel
    io = ExcelFile(
         ^^^^^^^^^^
  File "C:\python\venv\Lib\site-packages\pandas\io\excel\_base.py", line 1567, in __init__
    self._reader = self._engines[engine](
                   ^^^^^^^^^^^^^^^^^^^^^^
  File "C:\python\venv\Lib\site-packages\pandas\io\excel\_openpyxl.py", line 553, in __init__
    super().__init__(
  File "C:\python\venv\Lib\site-packages\pandas\io\excel\_base.py", line 573, in __init__
    self.book = self.load_workbook(self.handles.handle, engine_kwargs)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
TypeError: load_workbook() takes 2 positional arguments but 3 were given
```


Here is updated monkey-patching for pandas version 2.2.3, take attention on `# if you want ONLY hyperlink` comment:
P.S.: Works only with xlsx, i.e., engine is openpyxl
P.P.S.: If you reading this comment in the future and issue still Open, don't forget to see changes in `_convert_cell` and `load_workbook` at `pandas.io.excel._openpyxl` and update them accordingly.

```
from openpyxl import Workbook
import numpy as np
from pandas._typing import (
    FilePath,
    ReadBuffer,
    Scalar,
)


def _convert_cell(self, cell) -> Scalar:
    from openpyxl.cell.cell import (
        TYPE_ERROR,
        TYPE_NUMERIC,
    )
    if cell.hyperlink and cell.hyperlink.target:
        # return cell.hyperlink.target # if you want ONLY hyperlink
        return f'{cell.value}|||{cell.hyperlink.target}'  # if you want return value and hyperlink, separated by |||
    # here starts original code, except for "if" became "elif"
    elif cell.value is None:
        return ""  # compat with xlrd
    elif cell.data_type == TYPE_ERROR:
        return np.nan
    elif cell.data_type == TYPE_NUMERIC:
        val = int(cell.value)
        if val == cell.value:
            return val
        return float(cell.value)
    return cell.value


def load_workbook(self, filepath_or_buffer: FilePath | ReadBuffer[bytes], engine_kwargs) -> Workbook:
    from openpyxl import load_workbook
    default_kwargs = {"read_only": False, "data_only": True, "keep_links": False}
    return load_workbook(
        filepath_or_buffer,
        **(default_kwargs | engine_kwargs),
    )


OpenpyxlReader._convert_cell = _convert_cell
OpenpyxlReader.load_workbook = load_workbook

# after code above Excel file will have hyperlinks on openings
# parameters may vary depending on your code, i.e. just use this function as you want:
df = pandas.read_excel(excel_file, sheet_name=None, engine='openpyxl')
```

**Comment by moi90** [comment]
I propose to introduce `read_excel(..., hyperlinks="label")` as an explicit, engine-agnostic way to control how Excel hyperlinks are processed. The default `"label"` would preserve current behavior by reading the visible cell text, while `"destination"` could return the link destination URL.

> One possible enhancement is to allow user funcs (passed via `converters` kwds) to accept excel `Cell` instances.
> 
> @sinhrks

Is there an advantage of this more general solution? It would be completely engine-specific, which I would see as a downside of this approach.

**Comment by rhshadrach** [comment]
If hyperlinks could be extracted for all engines, I'd be +0.5 here. The two things keeping me from +1 are:

- Running with or without getting the links is dropping data, meaning that the user that wants both has to read the sheet twice which is not great. That said, I haven't been able to come up with a good API that resolves this.
- Links are just one example of things pandas drops. I'd be much more receptive to a solution could handle all additional data in a sheet.

That said, the fact that this can only be accomplished by openpyxl makes me quite negative here; more fragmentation in the Excel API between engines outweighs the benefits.

**Comment by moi90** [comment]
@rhshadrach 

> more fragmentation in the Excel API between engines outweighs the benefits.

I see what you mean. An alternative would be to just allow the user to provide a callable to interpret the raw data, like the workarounds above, but without the monkey-patching. It would be engine-specific, but the caller has control over the specific engine. What do you think about that?

**Comment by rhshadrach** [comment]
I think that's a good idea @moi90 - I'm supportive of it assuming there is a reasonable implementation.

**Comment by moi90** [comment]
@rhshadrach I propose the following, which would give a general mechanism useful beyond hyperlink extraction:

- Add a new `read_excel` parameter `convert_cell` and thread it through the full Excel read path (public API → base reader → concrete engines).
2. Each engine applies custom conversion per cell when provided. Current behavior stays unchanged when not provided.

One thing that I'd like to discuss: I think it makes sense to allow a user to fall back to the default implementation. Either by passing the default implementation as a second argument, or by returning something like `NotImplemented`:

1. `convert_cell(cell, default_convert)`: Can call the engine's `default_convert`, either as a fall-back at the end. Or even earlier and then still modify the value.
2. `convert_cell(cell)`: Can return `NotImplemented`. In that case, the cell is converted using the engine's `default_convert`.

I favor `convert_cell(cell, default_convert)`. For the example of hyperlinks which sparked this discussion, a user could call `default_convert` to normalize the cell value, while also extracting the hyperlink and return a merge of both in the end.

What do you think?

**Comment by rhshadrach** [comment]
Looking into this more, read_excel already has `converters` which is very close to this feature. I do not think we should have both. E.g. I think this code would be very confusing:

    pd.read_excel(..., convert_cell=..., converters=...)

Two options I see here:

1. Introduce `convert_cell` as proposed in https://github.com/pandas-dev/pandas/issues/13439#issuecomment-4646508925 with the signature `convert_cell(cell, default_convert)` (possibly renaming the argument), and deprecate `converters`. This would break the consistency with e.g. `read_csv` and code such as `converters={"a": lambda x: round(x, 2)}` would become more verbose.
2. Introduce something like `raw_converters` in parallel to `converters`. Here we could possibly expand the allowed arguments to be a callable similar to `convert_cell`.

One thing I think that would be good to answer before proceeding: why do users need `converters`? It seems to me users can apply converters to the result of `read_excel` or `read_csv`. Is it purely for convenience or does having this argument allow users to accomplish something they otherwise could not?

**Comment by moi90** [comment]
@rhshadrach The problem is that `converters` and `convert_cell` operate at different stages of processing:

1. `BaseExcelReader.get_sheet_data`: Convert raw Cell objects to Python datatypes (int, float, str, datetime, ...). Each Cell is converted by the reader's `_convert_cell`, e.g. `OpenpyxlReader._convert_cell`. (At this stage, there is only individual cells, no named columns etc.)
2. `BaseExcelReader.parse`: Collect rows returned from `get_sheet_data`. Ultimately, the data is passed to `pandas.io.parsers.TextParser` which applies `converters`, `dtype`, `parse_dates`, na settings, ... (Only after this step, the data has the shape of a DataFrame.)

That means:
- `convert_cell` is engine-level and can inspect rich cell metadata (for example a hyperlink target on openpyxl cells) before flattening. It is applied to *all* cells of a sheet, without being able to distinguish between different columns.
- `converters` is table/column-level and works on already-converted scalar values, aligned by column names and indices.

> One thing I think that would be good to answer before proceeding: why do users need `converters`? It seems to me users can apply converters to the result of `read_excel` or `read_csv`. Is it purely for convenience or does having this argument allow users to accomplish something they otherwise could not?

- `converters` is a parameter to `TextParser` (and therefore the same for all functions that rely on `TextParser`, e.g. `read_csv`).

- Parse-stage semantics: Converter functions are applied before type inference. That can change inferred dtypes and NA behavior. If you convert after reading, you operate on already-inferred columns.

- Converters let users sanitize problematic tokens as data is parsed, instead of first materializing a potentially problematic or bloated intermediate representation and then fixing it. (At first sight, `convert_cell` could serve the same purpose, but actually, it is different: Converter functions are applied to individual columns, `convert_cell` are not.)

---

I agree that it would be neat to be able to interpret raw cell instances in `converters` when we already know the column. Maybe one could use some kind of envelope object that holds the result of the respective reader's `_convert_cell`, but also keeps a reference to the original cell instance. I did some investigating, but I couldn't come up with an idea how to hide this envelope object from basically the whole `TextParser` machinery but then pass it to `converters`. Probably, this would require the introduction of an `class AnnotatedCell: value: Any; raw_value: Any | None = None` (or something like that) on the `TextParser` level, so that column name extraction (and so forth) could use the interpreted `value`, but `converters` receives the complete `AnnotatedCell` instance and, therefore, has access to the raw cell. However, this would break backward compatibility, so we would *still* need `converters` (receiving `AnnotatedCell.value`) *and* `raw_converters` (receiving the `AnnotatedCell` instance).

Overall, that would be a much larger change in a central place (`TextParser`) that I find harder to justify than my idea proposed earlier.

---

For the reasons outlined above, I would still go with `convert_cell` (as proposed earlier) and also keep `converters`.

**Comment by rhshadrach** [comment]
> [@rhshadrach](https://github.com/rhshadrach) The problem is that `converters` and `convert_cell` operate at different stages of processing:

I am indeed aware of that, my comment about being close is not that "adding would be of no value"; it's that having to similar arguments is a confusing API design. If it's not possible to avoid that's one thing, but I do not believe that to be the case here.

> * Parse-stage semantics: Converter functions are applied before type inference. That can change inferred dtypes and NA behavior. If you convert after reading, you operate on already-inferred columns.

Can't users disable inference by specifying `dtype=object`?

> * Converters let users sanitize problematic tokens as data is parsed, instead of first materializing a potentially problematic or bloated intermediate representation and then fixing it.

Isn't that only an illusion? The "potentially problematic or bloated intermediate representation" is still materialized, it's just happening within pandas.

> For the reasons outlined above, I would still go with `convert_cell` (as proposed earlier) and also keep `converters`.

I am quite negative on this approach.

**Comment by moi90** [comment]
> having t[w]o similar arguments is a confusing API design

I'll try three interpretations of your statement, as I'm not sure which one you meant... (I swear I'm not a bot, I'm just trying to understand your concerns :D)

### Having *two* similar arguments is a confusing API design

I agree, but I can not think of a way to make it work with just one argument.

- Let's suppose we drop `converters`. That would immediately break backward compatibility, so let's not do that.

This leaves us with these options:

- Implement the new functionality using `converters`. Making this work would require us to somehow hide the raw cell object from most of the `TextParser` (e.g. column header detection) until it is passed to the respective converter. For backwards-compatibility reasons, the converter must, by default, still receive the interpreted value. I see two options:

  1. Introduce a new parameter `pass_raw_cells` (or something like that) to instruct the reader to pass raw cell values. (Yet another parameter. Users might want to configure this individually per column.)
  2. Somehow hide the cell instance inside the interpreted value. (It feels brittle to just assume that we can just attach a `_raw_cell` attribute to any object.)

  I don't like neither option. Both require extensive changes to `TextParser`.

- Introduce something like `raw_converters` in parallel to `converters` (your suggestion).  Same problem as above: Requires extensive changes to the central `TextParser` component, although only `read_excel` requires it. Could work, but requires a lot of work.

- Leave `converters` untouched and add an additional argument to hook into the cell parsing stage, before the result is passed to `TextParser` (my suggestion). This change would only touch `read_excel`. To not make it confusing, explain the two stages of parsing as clear and with as few words as possible in the doc string. Still my favorite :)

### Having two *similarly named* arguments is a confusing API design

What about `[raw_]cell_parser`, `[raw_]cell_decoder` or something like that?

---

> having t[w]o similar arguments is a confusing API design. If it's not possible to avoid that's one thing, but I do not believe that to be the case here.

Well, can you *prove* that it is possible to avoid having these two arguments, e.g. by outlining a viable solution? I tried to follow your arguments, but I was not able to find a solution that leaves me happy...

---

One the "why do users need `converters`?" topic: I tried to give an answer, but I feel that, ultimately, this is not the place to discuss this... The parameter exists and it would be a breaking change to remove that parameter.

---

I've done my best to find a good solution here, but if I couldn't convince you, I'm going to let this matter rest. It probably needs a fresh pair of eyes, then.

**Comment by rhshadrach** [comment]
@moi90 - apologies on the delay here, it took several attempts but I think I have a proposal I like.

Add a `.links` attribute to `ExcleFile`: a dictionary mapping sheetname to the links of the file. They will be stored in a DataFrame with the same shape as the data, `NaN` when the cell has no link.

```python
with pd.ExcelFile("vendors.xlsx") as xl:
    df = xl.parse("Vendors")
    links = xl.links["Vendors"]
```

If a sheet is read multiple times, last one is stored in the dict. This avoids the two-pass issue (links is populated on parse). There is a little duplication when there is only one sheet name, but I think it's okay (this could also be reduced via `xl.sheet_names`).

One question: do we add a `capture_links=[True|False]` argument to `pd.ExcelFile` or just always read links when it's possible? I'd need to dig a bit more into what an implementation looks like to have a real opinion here, but I think I could be okay either way.

Thoughts?

**Comment by moi90** [comment]
@rhshadrach It's great that you found time to look into this, and thank you
for including me in your thought process!

I'm curious: How do you intend to construct the `links` DataFrame without touching `TextParser` and, at the same time, ensuring that `df` and `links` stay aligned? (It will become complicated as soon as parameters like `names`, `header`, `index_col` become involved and I still believe that `TextParser` will need to be altered to obtain a robust solution. Or do I overlook something obvious?)

How well would your solution support round-tripping? Excel supports hyperlinks both through `=HYPERLINK(...)`
formulas and as a cell property (as used by
[`[XlsxWriter.Worksheet.write_url()](https://xlsxwriter.readthedocs.io/worksheet.html#worksheet-write-url)`](https://xlsxwriter.readthedocs.io/worksheet.html#worksheet-write-url) or [`[openpyxl.Cell.hyperlink](https://openpyxl.readthedocs.io/en/stable/api/openpyxl.cell.cell.html#openpyxl.cell.cell.Cell.hyperlink)`](https://openpyxl.readthedocs.io/en/stable/api/openpyxl.cell.cell.html#openpyxl.cell.cell.Cell.hyperlink)).
With the current `DataFrame.to_excel` API, users can only reconstruct links through formulas:

```python
with pd.ExcelFile("vendors.xlsx", capture_links=True) as xl:
    df = xl.parse("Vendors")
    links = xl.links["Vendors"]

df["Vendor"] = [
    f'=HYPERLINK("{url}", "{label}")' if not pd.isna(url) else label
    for label, url in zip(df["Vendor"], links["Vendor"])
]

df.to_excel("vendors_with_links.xlsx", index=False)
```

That will often be sufficient, but it cannot faithfully preserve a property-based hyperlink. It also requires users to carry two aligned DataFrames and combine them manually.

> One question: do we add a `capture_links=[True|False]` argument to `pd.ExcelFile` or just always read links when it's possible?

I would make it opt-in: retrieving links may need engine-specific work and impose memory or performance costs.

---

There may also be a more general direction. Earlier you said that links are only one example of information pandas drops and that a solution which could handle additional sheet data would be more compelling. An attribute-based solution seems likely to produce further `.comments`, `.styles`, etc. attributes over time.

An alternative would be to introduce an intermediate cell representation (and expose it to the user). This would be a substantially larger design (and might require changes to `TextParser`, but your solution might as well), but it could provide a coherent way to expose hyperlinks, comments, and other metadata without committing to one special-purpose sidecar per property. It could also provide a foundation for true round-tripping of hyperlinks and other metadata if the Excel writers were extended accordingly to read this information. In fact, such an intermediate cell representation is already used in `pandas/io/formats/excel.py` to pass cell styles from `Styler` to the respective Excel writer.

(The ultimate test would be writing data to a specific location in an Excel file (skipping a number of rows), applying styles and "native" hyperlinks, using one of the Python Excel writer libraries. Then, read it back using Pandas (with the new "rich" representation) and write it again to a new Excel file, using the same row offset. In the end, both files should be mostly identical.)

---

(Sorry for the -once more- way too long message. I can't help but look for the best-possible solution here...)

**Comment by rhshadrach** [comment]
I'll have to think on this more, but my immediate reaction to the bottom of your comment about other attributes and round-tripping desires is "we should consider this issue and other things as outside the scope of pandas".

I'm not sure how my proposal on where we store the result will mean anything different about the implementation compared to the others that have been proposed. In other words, I don't see a reason why the API I've propose would present any more or less difficultly in implementation, and for that reason I choose to punt thinking more on implementation until we have an API we want to go after.

**Comment by moi90** [comment]
@rhshadrach

> I don't see a reason why the API I've propose would present any more or less difficultly in implementation, and for that reason I choose to punt thinking more on implementation until we have an API we want to go after.

True! That is also my conclusion.

My main criterion for the API is that it should be straightforward and intuitive to jointly process a cell's URL and text. Your proposed solution supports this requirement. (It also seems to me that we can not find something better than a set of distinct DataFrames for the different attributes.) I'm not entirely happy with the asymmetry of your solution (`values = xl.parse` vs. `links = xl.links["Vendors"]`) and with the fact that additional metadata is accessed through dicts and attributes.

What do you think about something like this?

```python
with pd.ExcelFile("vendors.xlsx") as xl:
    sheet = xl.parse_with_metadata("Vendors")
    values = sheet.values # cell values, identical to the result of xl.parse
    hyperlinks = sheet.hyperlinks # cell hyperlinks
    # More attributes can be added naturally, e.g.:
    styles = sheet.styles # cell styles
```

`sheet` is a `namedtuple` whose fields are aligned DataFrames, e.g. values and links.

**Comment by rhshadrach** [comment]
> An attribute-based solution seems likely to produce further .comments, .styles, etc. attributes over time.

This comment made me rethink whether or not hyperlink extraction belongs in pandas. I'm now of the opinion that extracting anything but the values in the sheets should be outside the scope of pandas and is more appropriate in Excel-reader specific packages.

One possible criticism here is that a 3rd party could not match behavior nor be as performant because pandas can use internals. From what I can tell, this does not happen in the implementation. While Excel/parser does make calls to private methods, those that are relevant are already available in the public API.

On top of this is that this would further introduce divergent behaviors between engines. But even without this, I'd still be quite negative here.

cc @jorisvandenbossche @jbrockmendel @mroeschke

**Comment by jbrockmendel** [comment]
I don't feel strongly about this, but lean toward rhshadrach's position to consider this out of scope.

**Comment by moi90** [comment]
> One possible criticism here is that a 3rd party could not match behavior nor be as performant because pandas can use internals. From what I can tell, this does not happen in the implementation.

Yes, that is true! It is my understanding that even `TextParser` itself is not public API. Am I mistaken?

But I agree: It might be warranted to solve this externally. (😭)
